kubemarine/core/executor.py

```diff
@@ -118,288 +133,206 @@ def reparse_results(self, results, batch):
-        if not batched_results:
-            return
-        output = ""
-        for conn, results in batched_results.items():
-            for token, result in results.items():
-                if isinstance(result, invoke.exceptions.UnexpectedExit):
-                    result = result.result
-
-                # for now we do not know how-to print transfer result
-                if not isinstance(result, fabric.runners.Result):
-                    continue
-
-                if output != "":
-                    output += "\n"
-                output += "\t%s (%s): code=%i" % (conn.host, token, result.exited)
-                if result.stdout:
-                    output += "\n\t\tSTDOUT: %s" % result.stdout.replace("\n", "\n\t\t        ")
-                if result.stderr:
-                    output += "\n\t\tSTDERR: %s" % result.stderr.replace("\n", "\n\t\t        ")
-
-        return output
-
-    def print_last_results(self) -> None:
-        """
-        Prints debug message to log with last executed queued command results.
-        :return: None
+    def flush(self) -> Dict[str, TokenizedResult]:
         """
-        results = self.get_last_results_str()
-        self.cluster.log.debug(results)
+        Flushes the connections' queue and returns grouped result
 
-    def flush(self) -> dict:
-        executor = self._get_active_executor()
-
-        batch_results = {}
+        :return: grouped tokenized results per connection.
+        """
+        batch_results: Dict[str, TokenizedResult] = {}
 
-        if not executor.connections_queue:
-            executor.cluster.log.verbose('Queue is empty, nothing to perform')
+        if not self._connections_queue:
+            self.logger.verbose('Queue is empty, nothing to perform')
+            self._last_results = batch_results
             return batch_results
 
-        callable_batches = executor._get_callables()
+        callable_batches: List[Dict[str, _MergedPayload]] = self._get_callables()
 
-        max_workers = len(executor.connections_queue.keys())
-        if not executor.parallel:
-            max_workers = 1
+        max_workers = len(self._connections_queue)
 
         with ThreadPoolExecutor(max_workers=max_workers) as TPE:
+            failed_hosts: Set[str] = set()
             for batch in callable_batches:
-                results = {}
-                futures = {}
+                results: _RawHostToResult = {}
+                futures: Dict[str, concurrent.futures.Future] = {}
 
-                def safe_exec(result_map: Dict[str, Any], key: str, call: Callable[[], Any]):
+                def safe_exec(result_map: Dict[str, Any], host: str, call: Callable[[], Any]) -> None:
                     try:
-                        # sleep required to avoid thread starvation
-                        time.sleep(0.1)
-                        result_map[key] = call()
-                        time.sleep(0.1)
+                        result_map[host] = call()
```

Comment:
time.sleep was an unreliable workaround to fix bug in async transferring of files. The bug is fixed in https://github.com/Netcracker/KubeMarine/commit/815df335fd697485ac0d3a63f8eaee23c792979c